[ROCm] Add amdgpu_arch string as member variable in ir_emitter_context

deven-amd opened the pull request:
This PR is a pre-req for a follow-up PR which will introduce gfx arch specific codegen in the IR emitter code.

-------------------------

/cc @chsigg @cheshire

cheshire commented on tensorflow/compiler/xla/service/gpu/gpu_compiler.cc:
```diff
@@ -852,7 +852,7 @@ static Status CompileModuleToLlvmIrImpl(
     const std::string& target_triple, const std::string& data_layout,
     const std::string& platform_name, GpuDeviceInfo gpu_device_info,
-    se::CudaComputeCapability cuda_compute_capability,
+    se::CudaComputeCapability cuda_compute_capability, std::string amdgpu_arch,
```
Maybe absl::string_view?

cheshire replied:
I guess if we pass CudaComputeCapability it might make sense to pass AMD capabilities as well.

Two questions though:

1) Maybe it's better to have a class instead?
2) What about passing `absl::variant<>` of both? While we're there, can create a wrapping class GpuComputeCapability.

deven-amd replied:
@cheshire , are you proposing that we should
* create a `se::RocmComputeCapability` class 
* replace references to `CudaComputeCapability` with references to `absl::variant<RocmComputeCapability, CudaComputeCapability>` instead?

There are a lot of references to `CudaComputeCapability` in the codebase, will all of them need to be updated to use the `absl::variant<...>` ?

Note that having a common `GpuComputeCapability` abstract class to define an interface that is then implemented by concrete `RocmCompuateCapability`/`CudaComputeCapability` class might not work from a design point of view, since the two do not share anything in common. There is `major` / `minor` versioning with `gfx...` HW, and feature support (fast fp16, fast matrix math, etc) is determined on a per `gfx...` type basis.

deven-amd replied:
@cheshire ,  please review the change in the following PR (in the ROCm fork)
https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/pull/1595

It introduces the `RocmComputeCapability` wrapper that you have proposed here. Once we have it integrated in the ROCm fork, I can update this PR to include those changes

thanks

gbaned commented:
@deven-amd Can you please check @cheshire's comments and keep us posted ? Thanks!

deven-amd committed:
Adding a `rocm_compute_capacbility` field to the `DeviceDescription` class

Currently, in the `DeviceDeescription` class,  we store the
* `rocm_amdgpu_isa_version` as an int member
* `rocm_amdgpu_gcn_arch_name` as a string member

The `rocm_amdgpu_isa_version` int member usage, which stores integer suffix in the gfx arch (i.e 900 from "gfx900") is already flawed with the introduction of "gfx90a", and needs to dropped

The `rocm_amdgpu_gcn_arch_name` string usages needs to be updated, to make it similar to what is done on the CUDA side.

This commit

* introduces a class `RocmComputeCapability`, that wraps the `gcn_arch_name` routine and provides helpers functions to do all the `gcn_arch_name` based checks, that are currently dispersed in the TF code.

* adds an instance of the `RocmCompuateCapability` class, `rocm_compute_capability`, as a member of `DeviceDescription`

There will be follow-up commits to remove the `rocm_amdgpu_isa_version` and `rocm_amdgpu_gcn_arch_name` members from the `DeviceDescription` class and replace their usage with `rocm_compute_capability` instead

deven-amd committed:
Removing a `rocm_amdgpu_isa_version` field from the `DeviceDescription` class

This commit removes the `rocm_amdgpu_isa_version` member from the `DeviceDescription` class and replace its usage with `rocm_compute_capability` instead

deven-amd committed:
Removing a `rocm_amdgpu_gcn_arch_name` field from the `DeviceDescription` class

This commit removes the `rocm_amdgpu_gcn_arch_name` member from the `DeviceDescription` class and replace its usage with `rocm_compute_capability` instead

deven-amd committed:
Replacing in XLA code, references to `amdgpu_arch` string with rerferences to `rocm_compute_capability` instead

deven-amd commented:
@cheshire, please re-review....I have updated this branch with a subset of the changes you reviewed for the PR to introduce `RocComputeCapability` in the rocm fork.

Also I see that the `Code Check - Changed Files` is failing, but am unable to determine what exactly is causing the failure...please let me know if that is something I need to look into

cheshire commented on tensorflow/compiler/xla/service/gpu/amdgpu_compiler.cc:
```diff
@@ -119,14 +119,7 @@ AMDGPUCompiler::AMDGPUCompiler()
-  std::string gcn_arch_name =
-      stream_exec->GetDeviceDescription().rocm_amdgpu_gcn_arch_name();
-  if (gcn_arch_name == stream_exec->GetDeviceDescription().kUndefinedString) {
-    LOG(WARNING) << "Couldn't get AMDGPU GCN Arch for device; assuming gfx900.";
```
Do we know when and why this could happen? In general I'm not keen on ignoring errors with logs.

deven-amd replied:
This should not happen at all, in theory. The check here is just something we had in place for a long time...we can get rid of it, or convert to error, if need be

gbaned commented:
@deven-amd  Can you please check @cheshire's comments and keep us posted ? Thanks!

deven-amd commented:
@gbaned , thought @cheshire already approved the PR. Please merge....there are few other PRs we need to file, but are dependent on the changes in this PR.

thanks

deven-amd commented:
@gbaned anything I can do on my end to get the PR merged? thanks

gbaned commented:
> @gbaned anything I can do on my end to get the PR merged? thanks

@deven-amd  This PR is processing internally, nothing pending from your end. Thank you.
